apollo-router/src/plugins/traffic_shaping/deduplication.rs

```diff
@@ -124,35 +124,42 @@ where
-                    let res = {
+                    let (res, handle) = {
                         // when _drop_signal is dropped, either by getting out of the block, returning
                         // the error from ready_oneshot or by cancellation, the drop_sentinel future will
                         // return with Err(), then we remove the entry from the wait map
                         let (_drop_signal, drop_sentinel) = oneshot::channel::<()>();
-                        tokio::task::spawn(async move {
+                        let handle = tokio::task::spawn(async move {
                             let _ = drop_sentinel.await;
                             let mut locked_wait_map = wait_map.lock().await;
                             locked_wait_map.remove(&cache_key);
                         });
 
-                        service.call(request).await.map(CloneSubgraphResponse)
+                        (
+                            service.call(request).await.map(CloneSubgraphResponse),
+                            handle,
+                        )
                     };
 
-                    // Let our waiters know
-
-                    // Clippy is wrong, the suggestion adds a useless clone of the error
-                    #[allow(clippy::useless_asref)]
-                    let broadcast_value = res
-                        .as_ref()
-                        .map(|response| response.clone())
-                        .map_err(|e| e.to_string());
+                    // Make sure that our spawned task has completed. Ignore the result to preserve
+                    // existing behaviour.
+                    let _ = handle.await;
+                    // At this point we have removed ourselves from the wait_map, so we won't get
+                    // any more receivers. If we have any receivers, let them know
+                    if tx.receiver_count() > 0 {
+                        // Clippy is wrong, the suggestion adds a useless clone of the error
+                        #[allow(clippy::useless_asref)]
+                        let broadcast_value = res
+                            .as_ref()
+                            .map(|response| response.clone())
+                            .map_err(|e: &BoxError| e.to_string());
 
-                    // We may get errors here, for instance if a task is cancelled,
-                    // so just ignore the result of send
-                    let _ = tokio::task::spawn_blocking(move || {
-                        tx.send(broadcast_value)
-                    }).await
-                    .expect("can only fail if the task is aborted or if the internal code panics, neither is possible here; qed");
+                        // We may get errors here, for instance if a task is cancelled,
+                        // so just ignore the result of send
+                        let _ = tokio::task::spawn_blocking(move || {
+                            tx.send(broadcast_value)
```

Comment:
I'm actually not sure why this is even in a spawn_blocking. `send()` does access a mutex but so does `tx.receiver_count()`. I think it's unnecessary

Reply:
I don't like changing code that isn't part of the fix is the reason why I left it and I think the motivation was to prevent the send being lost if the task was cancelled. However, the loop takes care of that for us and so I think you are right and we can remove the spawn_blocking. I'll do that. Fixed: https://github.com/apollographql/router/pull/7347/commits/f9590bcf3f82910226e94ef122d43d9d92cb211a